File: TsinghuaNetBackground/LiveTileTask.cpp
#include "pch.h"

#include "LiveTileTask.h"

using namespace winrt;
using namespace Windows::ApplicationModel::Background;
using namespace Windows::Foundation;
using namespace TsinghuaNetHelper;

namespace winrt::TsinghuaNetBackground::implementation
{
    NetState GetSuggestNetState(SettingsHelper const& lan)
    {
        hstring ssid;
        auto status = lan.GetCurrentInternetStatus(ssid);
        switch (status)
        {
        case InternetStatus::Lan:
            return lan.LanState();
        case InternetStatus::Wwan:
            return lan.WwanState();
        case InternetStatus::Wlan:
            return lan.WlanState(ssid);
        default:
            return NetState::Unknown;
        }
    }

    IConnect GetHelper(NetState state)
    {
        switch (state)
        {
        case NetState::Auth4:
            return Auth4Helper();
        case NetState::Auth6:
            return Auth6Helper();
        case NetState::Net:
            return NetHelper();
        default:
            return nullptr;
        }
    }

    IAsyncAction LiveTileTask::Run(IBackgroundTaskInstance const& taskInstance)
    {
        auto deferral = taskInstance.GetDeferral();
        try
        {
            NetState state = GetSuggestNetState(settings);
            IConnect helper = GetHelper(state);
            if (helper)
            {
                FluxUser user = co_await helper.FluxAsync();
                notification.UpdateTile(user);
            }
        }
        catch (hresult_error const&)
        {
        }
        deferral.Complete();
    }
} // namespace winrt::TsinghuaNetBackground::implementation

File: TsinghuaNet.XF/TsinghuaNet.XF.UWP.Background/LiveTileTask.cs
using TsinghuaNet.Models;
using TsinghuaNet.XF.Models;
using TsinghuaNet.XF.UWP.Helpers;
using Windows.ApplicationModel.Background;

namespace TsinghuaNet.XF.UWP.Background
{
    public sealed class LiveTileTask : IBackgroundTask
    {
        public async void Run(IBackgroundTaskInstance taskInstance)
        {
            var deferral = taskInstance.GetDeferral();
            try
            {
                NetCredential credential = new NetCredential();
                InternetStatus status = new InternetStatus();
                credential.State = await status.SuggestAsync();
                NetXFSettings settings = new NetXFSettings();
                settings.LoadSettings();
                credential.UseProxy = settings.UseProxy;
                var helper = credential.GetHelper();
                if (helper != null)
                {
                    FluxUser user = await helper.GetFluxAsync();
                    NotificationHelper.UpdateTile(user);
                    if (settings.EnableFluxLimit)
                        NotificationHelper.SendWarningToast(user, settings.FluxLimit);
                }
            }
            finally
            {
                deferral.Complete();
            }
        }
    }
}
